### PR TITLE
Adds a dev guide section on Rust Coverage

### PR DESCRIPTION
**Pull request by richkadel**

### CODE REVIEW

**Comment by richkadel** [comment]
cc: @tmandry @wesleywiser

**Comment by richkadel** [comment]
Dev guide section for tracking issue rust-lang/rust#79121

File: src/llvm-coverage-instrumentation.md [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+physical counters (or any other executable instructions) in the final binary.
+Some of them will be (see `CoverageKind::`[`Counter`][counter-coverage-kind]),
+but other counters can be computed on the fly, when generating a coverage
+report, by mapping a `CodeRegion` to a
+`CoverageKind`::[`Expression`][expression-coverage-kind].
+
+As an example:
+
+```rust
+fn some_func(bool flag) {
```
Comment (jyn514) [inline src/llvm-coverage-instrumentation.md:93]:
```suggestion
fn some_func(flag: bool) {
```

File: src/llvm-coverage-instrumentation.md [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+function, generic, or closure), the `Instrumentor`'s constructor prepares a
+[`CoverageGraph`][coverage-graph] and then executes
+[`inject_counters()`][inject-counters].
+
+```rust
+        Instrumentor::new(&self.name(), tcx, mir_body).inject_counters();
+```
+
+The `CoverageGraph` is a coverage-specific simplification of the MIR control
+flow graph (CFG). It's nodes are [`BasicCoverageBlock`s][bcb], which
```
Comment (jyn514) [inline src/llvm-coverage-instrumentation.md:314]:
```suggestion
flow graph (CFG). Its nodes are [`BasicCoverageBlock`s][bcb], which
```

**Review by jyn514 — COMMENTED** [review]
Wow, this is super detailed :smile: I'll let tmandry review the content, I just have some typographical nits.

File: src/llvm-coverage-instrumentation.md (hunk 1 above)
Comment (richkadel) [reply src/llvm-coverage-instrumentation.md]: fixed

File: src/llvm-coverage-instrumentation.md (hunk 2 above)
Comment (richkadel) [reply src/llvm-coverage-instrumentation.md]: fixed

**Comment by richkadel** [comment]
> typographical nits

@jyn514 - Thanks for finding the typos!

**Comment by richkadel** [comment]
@tmandry - went through all of the links and corrected issues. I think the links are all valid now.

File: src/llvm-coverage-instrumentation.md [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+
+[compile-codegen-unit]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/base/fn.compile_codegen_unit.html
+[coverageinfo-finalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/context/struct.CodegenCx.html#method.coverageinfo_finalize
+[mapgen-finalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/mapgen/fn.finalize.html
+[coverage-mapping-format]: https://llvm.org/docs/CoverageMappingFormat.html
+[add-unreachable-coverage]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/mapgen/fn.add_unreachable_coverage.html
+
+## Testing LLVM Coverage
+
+Coverage instrumentation in the MIR is validated by `mir-opt` test:
```
Comment (wesleywiser) [inline src/llvm-coverage-instrumentation.md:268]:
```suggestion
Coverage instrumentation in the MIR is validated by a `mir-opt` test:
```

**Comment by wesleywiser** [comment]
It might be good to have a small discussion about the code coverage format version: Newer versions generally have features we might like to use but aren't supported by the min version of LLVM rustc allows. If the version of LLVM is too old, we generate a compiler error when using the code coverage feature. 

Overall, this is really great work! 👍

File: src/llvm-coverage-instrumentation.md (hunk 3 above)
Comment (richkadel) [reply src/llvm-coverage-instrumentation.md]: Done

**Comment by richkadel** [comment]
> It might be good to have a small discussion about the code coverage format version: Newer versions generally have features we might like to use but aren't supported by the min version of LLVM rustc allows. If the version of LLVM is too old, we generate a compiler error when using the code coverage feature.

@wesleywiser - Added as a footnote. PTAL.

 Thanks!

**Review by wesleywiser — APPROVED** [review]
Looks great!

**Pull request merged**